tests/README.md

```diff
@@ -0,0 +1,1510 @@
+$_POST["username"] = 'admin';
+$_GET["client_id"] = 'default';
+```
+
+The values should be set accordingly to the ILIAS setup.
+
+<a name="exclude-ilias-bound-tests"></a>
+#### Exclude ILIAS bound tests
+- Navigate to "Run -> Edit Configurations..."
+- Select your PHPUnit run configuration
+- Add "--exclude-group needInstalledILIAS" to the Test runner options.
+- Hit OK to save the changes
+
+<a name="run-only-ilias-bound-tests"></a>
+#### Run only ILIAS bound tests
+- Navigate to "Run -> Edit Configurations..."
+- Select your PHPUnit run configuration
+- Add "--group needsInstalledILIAS" to the Test runner options.
+- Hit OK to save the changes
+
+<a name="run-tests-with-cli"></a>
+### Run tests with CLI
+At the current state of the ILIAS test, these require the backup of the global
+scope which indicates a dependency between some of the tests. Therefore, if the
+test should run as configured on the CI server omit the *--no-globals-backup*,
+*--report-useless-tests* and *--disallow-todo-tests* options.
+
+The commands bellow must be run from the ILIAS web root directory.
+
+<a name="execute-ilias-installation-bound-tests"></a>
+#### Execute ILIAS installation bound tests
+```
+./libs/composer/vendor/bin/phpunit -c ./Services/PHPUnit/config/PhpUnitConfig.xml \
+	--colors=always \
+	--no-globals-backup \
+	--report-useless-tests \
+	--disallow-todo-tests \
+	--group needsInstalledILIAS
+```
+
+<a name="execute-all-tests"></a>
+#### Execute all tests
+```
+./libs/composer/vendor/bin/phpunit -c ./Services/PHPUnit/config/PhpUnitConfig.xml \
+	--colors=always \
+	--report-useless-tests \
+	--disallow-todo-tests \
+	--no-globals-backup
+```
+
+<a name="execute-only-installation-unbound-tests"></a>
+#### Execute only installation unbound tests
+```
+./libs/composer/vendor/bin/phpunit -c ./Services/PHPUnit/config/PhpUnitConfig.xml \
+	--colors=always \
+	--no-globals-backup \
+	--report-useless-tests \
+	--disallow-todo-tests \
+	--exclude-group needsInstalledILIAS
+```
+
+<a name="guidelines"></a>
+## Guidelines
+
+<a name="foreword"></a>
+### Foreword
+May the reader is asking him self why this guidelines refer to a book
+which writes about JUnit testing in Java 8.
+
+The reason for this is that unit testing in his very nature is the same in
+every language. The second reason is that no recently written books are
+on the market which describe modern unit testing with PHPUnit.
+
+<a name="naming"></a>
+### Naming
+"Rework test names and code to tell stories." (Langr 2015)
+This means to treat the tests as a specification which tells everything
+about the behaviours of a class.
+
+<a name="class"></a>
+#### Class
+The filename of the test class should always be named like *\<class name of the implementation\>Test.php*.
+Furthermore, the test class should always be named as the class which is tested by the unit test class.
+For example the real class is called *Car* the corresponding test class would be named *CarTest* and the filename
+ *CarTest.php*.
```

Comment:
Imo there also are reasonable definitions of "unit" that do not coincide with single classes. What to do with these?

Reply:
I will add a short chapter about namespaces.

Could you give me an example of such a unit ?  I am only asking because I never really encountered such a case :)

Reply:
Have a look at `\ILIAS\Data\Result`. The implementation in two classes `Ok` and `Error` is non-essential, what needs to be tested IMO is that the two factory methods `ok` and `error` do indeed return an object that behaves as specified in the `Result`-interface. Does this make any sense?